packages/crypto/src/bls12_318/pairing.rs

```diff
@@ -0,0 +1,234 @@
+use crate::{
+    errors::AggregationPairingEquality, CryptoError, BLS12_381_G1_POINT_LEN, BLS12_381_G2_POINT_LEN,
+};
+
+use super::points::{g1_from_variable, g2_from_variable};
+use ark_bls12_381::Bls12_381;
+use ark_ec::{
+    bls12::{G1Prepared, G2Prepared},
+    pairing::Pairing,
+};
+use num_traits::Zero;
+use rayon::iter::{ParallelBridge, ParallelIterator};
+
+pub fn bls12_381_aggregate_pairing_equality(
+    ps: &[u8],
+    qs: &[u8],
+    r: &[u8],
+    s: &[u8],
+) -> Result<bool, CryptoError> {
+    if ps.is_empty() {
+        return Err(AggregationPairingEquality::EmptyG1.into());
+    } else if qs.is_empty() {
+        return Err(AggregationPairingEquality::EmptyG2.into());
+    } else if ps.len() % BLS12_381_G1_POINT_LEN != 0 {
+        return Err(AggregationPairingEquality::NotMultipleG1 {
+            remainder: ps.len() % BLS12_381_G1_POINT_LEN,
+        }
+        .into());
+    } else if qs.len() % BLS12_381_G2_POINT_LEN != 0 {
+        return Err(AggregationPairingEquality::NotMultipleG2 {
+            remainder: qs.len() % BLS12_381_G2_POINT_LEN,
+        }
+        .into());
+    } else if (ps.len() / BLS12_381_G1_POINT_LEN) != (qs.len() / BLS12_381_G2_POINT_LEN) {
+        return Err(AggregationPairingEquality::UnequalPointAmount {
+            left: ps.len() / BLS12_381_G1_POINT_LEN,
+            right: qs.len() / BLS12_381_G2_POINT_LEN,
+        }
+        .into());
+    }
+
+    let pq_pairs: Vec<_> = ps
+        .chunks_exact(BLS12_381_G1_POINT_LEN)
+        .zip(qs.chunks_exact(BLS12_381_G2_POINT_LEN))
+        // From here on parallelism is fine since the miller loop runs over
+        // a sum of the pairings and is therefore a commutative operation
+        .par_bridge()
```

Comment:
I think we run into a problem here with determinism if there are multiple errors in the list. Then `.collect::<Result<_, CryptoError>>()?;` can lead to different errors.

Simple way out would be: we can compute this in parallel but must not change the order. Another option is to map all errors into the same error case. Thoughts?

Reply:
I'm thinking, if an intermediate allocation would be fine, we could first do the conversion into the structs of the elements (i.e. `g1_from_variable`) sequentially and then do the `G1Prepared::from(g1.0)` in parallel since that's doing some heavier computation.

That way we would keep most of the performance improvements while keeping error determinism.

Reply:
Sounds good. But what's the advantage of using `par_bridge()` over the parallel map function (don't recall the name) that also executes in parallel but preserves performance?

Reply:
There isn't really an advantage that I use it for, it's because I completely overlooked that rayon offers a `zip` on its indexed parallel iterators.

Maybe you know, does rayon guarantee that when we use `slice.into_par_iter().zip(slice2)`, that the pairs are always `(slice[0], slice2[0]), (slice[1], slice2[1]), ..)`?  
Because that's obviously an important property for the aggregation.

Reply:
It seems like this bridge is the issue here. It allows bridging all existing iterators, but loses the iteration order. https://docs.rs/rayon/latest/rayon/iter/trait.ParallelBridge.html

However, we can rewrite the whole thing with rayon from the very beginning. This then uses [IndexedParallelIterator](https://docs.rs/rayon/latest/rayon/iter/trait.IndexedParallelIterator.html#):

```diff
diff --git a/packages/crypto/src/bls12_318/pairing.rs b/packages/crypto/src/bls12_318/pairing.rs
index 3ef6b2ac4..03cb86589 100644
--- a/packages/crypto/src/bls12_318/pairing.rs
+++ b/packages/crypto/src/bls12_318/pairing.rs
@@ -9,7 +9,8 @@ use ark_ec::{
     pairing::Pairing,
 };
 use num_traits::Zero;
-use rayon::iter::{ParallelBridge, ParallelIterator};
+use rayon::iter::{IndexedParallelIterator, ParallelIterator};
+use rayon::slice::ParallelSlice;
 
 pub fn bls12_381_aggregate_pairing_equality(
     ps: &[u8],
@@ -40,11 +41,8 @@ pub fn bls12_381_aggregate_pairing_equality(
     }
 
     let pq_pairs: Vec<_> = ps
-        .chunks_exact(BLS12_381_G1_POINT_LEN)
-        .zip(qs.chunks_exact(BLS12_381_G2_POINT_LEN))
-        // From here on parallelism is fine since the miller loop runs over
-        // a sum of the pairings and is therefore a commutative operation
-        .par_bridge()
+        .par_chunks_exact(BLS12_381_G1_POINT_LEN)
+        .zip(qs.par_chunks_exact(BLS12_381_G2_POINT_LEN))
         .map(|(p, q)| {
             let g1 = g1_from_variable(p)?;
             let g2 = g2_from_variable(q)?;
```

Reply:
Okay, so I got rid of the `par_bridge` usage and played around a bit with the parallel iterators to force the negated `r` element into the parallel iterator as well and it seemed to work just fine, the tests still pass.